src/middlewared/middlewared/plugins/cluster_linux/ctdb_shared_volume.py

```diff
@@ -0,0 +1,213 @@
+            elif vol[0]['status'] != 'Started':
+                exists = True
+                await(
+                    await self.middleware.call('gluster.volume.start', self.CTDB_VOL_NAME)
+                ).wait(raise_error=True)
+                started = True
+            else:
+                exists = started = True
+
+        return exists, started
+
+    @accepts()
+    @job(lock=CRE_OR_DEL_LOCK)
+    async def create(self, job):
+        """
+        Create and mount the shared volume to be used
+        by ctdb daemon.
+        """
+
+        # check if ctdb shared volume already exists and started
+        exists, started = await self.shared_volume_exists_and_started()
+        if exists and started:
+            return
+
+        if not exists:
+            # get the peers in the TSP
+            peers = await(
+                await self.middleware.call('gluster.peer.pool')
+            ).wait(raise_error=True)
+            if not peers:
+                raise CallError('No peers detected')
+
+            # shared storage volume requires 3 nodes, minimally, to
+            # prevent the dreaded split-brain
+            con_peers = [i['hostname'] for i in peers if i['connected'] == 'Connected']
+            if len(con_peers) < 3:
+                raise CallError(
+                    '3 peers must be present and connected before the ctdb '
+                    'shared volume can be created.'
+                )
+
+            # create the ctdb shared volume
+            req = await self.construct_gluster_ctdb_api_request(con_peers)
+            await(
+                await self.middleware.call('gluster.volume.create', req)
+            ).wait(raise_error=True)
+
+        if not started:
+            # start it if we get here
+            await(
+                await self.middleware.call('gluster.volume.start', self.CTDB_VOL_NAME)
+            ).wait(raise_error=True)
+
+        # try to mount it locally
+        await(
+            await self.middleware.call('ctdb.shared.volume.mount')
+        ).wait(raise_error=True)
+
+        return 'SUCCESS'
+
+    @accepts()
+    @job(lock=CRE_OR_DEL_LOCK)
+    async def delete(self, job):
+        """
+        Delete and unmount the shared volume used by ctdb daemon.
+        """
+
+        # nothing to delete if it doesn't exist
+        exists, started = await self.shared_volume_exists_and_started()
+        if not exists:
+            return
+
+        # umount it first
+        await(
+            await self.middleware.call('ctdb.shared.volume.umount')
+        ).wait(raise_error=True)
+
+        if started:
+            # stop the volume
+            force = {'force': True}
+            await(
+                await self.middleware.call('gluster.volume.stop', self.CTDB_VOL_NAME, force)
+            ).wait(raise_error=True)
+
+        # now delete it
+        data = {'name': self.CTDB_VOL_NAME}
+        await(
+            await self.middleware.call('gluster.volume.delete', data)
```

Comment:
Is gluster.volume a CRUD or a simple Service?

Reply:
It's a CRUD service.

Reply:
Then I am confused. Following CRUD methods, delete method should receive an id/name as the first param, here we are providing a dict

Reply:
Because we're not storing cluster information in the db. I can go back and change the volume and peer plugins to SimpleService

Reply:
If these are not regular usual CRUD Services then it should not be a CRUD Service to begin with. I agree.